### ISSUE TITLE
CustomIntegrator Energy Issue

### ISSUE THREAD

**Issue by kyleabeauchamp** [pandegroup/openmm#880]
On the CUDA platform I'm finding that the potential energy stored in a custom integrator global variable is not agreeing with the value obtained by probing the state object.  Things work fine on the CPU platform.  Driver code is below:

```
import numpy as np
import simtk.openmm as mm
from simtk import unit as u
import simtk.openmm.app as app

temperature = 300. * u.kelvin

pdb = app.PDBFile("/home/kyleb/src/openmm/openmm/wrappers/python/simtk/openmm/app/data/tip3p.pdb")
forcefield = app.ForceField('tip3p.xml')

system = forcefield.createSystem(pdb.topology, nonbondedMethod=app.PME, nonbondedCutoff=1.0*u.nanometers)
positions = pdb.positions


integrator = mm.LangevinIntegrator(temperature, 1.0 / u.picoseconds, 0.25 * u.femtoseconds)
context = mm.Context(system, integrator)
context.setPositions(positions)
context.setVelocitiesToTemperature(temperature)

integrator.step(500)
positions = context.getState(getPositions=True).getPositions()
state = context.getState(getEnergy=True)
energy = state.getPotentialEnergy() + state.getKineticEnergy()
energy, state.getPotentialEnergy(), state.getKineticEnergy()


integrator = mm.CustomIntegrator(0.0005)
integrator.addGlobalVariable("U", 0.0)
integrator.addGlobalVariable("K", 0.0)
integrator.addPerDofVariable("x1", 0)
integrator.addUpdateContextState()
integrator.addComputePerDof("v", "v+0.5*dt*f/m")
integrator.addComputePerDof("x", "x+dt*v")
integrator.addComputePerDof("x1", "x")
integrator.addConstrainPositions()
integrator.addComputePerDof("v", "v+0.5*dt*f/m+(x-x1)/dt")
integrator.addConstrainVelocities()
integrator.addComputeGlobal("U", "energy")
integrator.addComputeSum("K", "0.5*m*v*v")

context = mm.Context(system, integrator)
context.setPositions(positions)
context.setVelocitiesToTemperature(temperature)


integrator.step(1)
positions = context.getState(getPositions=True).getPositions()
state = context.getState(getEnergy=True)
energy = state.getPotentialEnergy() + state.getKineticEnergy()
energy, state.getPotentialEnergy(), state.getKineticEnergy()

integrator.getGlobalVariableByName("U")
integrator.getGlobalVariableByName("K")
```

On CUDA the final output is this:

```
In [39]: energy, state.getPotentialEnergy(), state.getKineticEnergy()
Out[39]: 
(Quantity(value=-29064.035512298084, unit=kilojoule/mole),
 Quantity(value=-35961.64684042308, unit=kilojoule/mole),
 Quantity(value=6897.611328125, unit=kilojoule/mole))

In [40]: 

In [40]: integrator.getGlobalVariableByName("U")
Out[40]: 156570.0

In [41]: integrator.getGlobalVariableByName("K")
Out[41]: 6897.611328125

```

On CPU the final output is this:

```
In [39]: energy, state.getPotentialEnergy(), state.getKineticEnergy()
Out[39]: 
(Quantity(value=-29157.190375937946, unit=kilojoule/mole),
 Quantity(value=-35859.676374572875, unit=kilojoule/mole),
 Quantity(value=6702.485998634928, unit=kilojoule/mole))

In [40]: 

In [40]: integrator.getGlobalVariableByName("U")
Out[40]: -35859.676374572875

In [41]: integrator.getGlobalVariableByName("K")
Out[41]: 6702.485998634928



```

**Comment by rmcgibbo** [comment]
When you run these dynamics for a few more steps on CUDA, does it blow up or run correctly?

**Comment by kyleabeauchamp** [comment]
Correctly

**Comment by kyleabeauchamp** [comment]
I should also add that the same behaviour occurs with several other systems that I've tried, including alanine dipeptide in explicit solvent.

**Comment by kyleabeauchamp** [comment]
The energy mismatch also occurs on OpenCL.

**Comment by kyleabeauchamp** [comment]
Here's a simpler driver script that still induces the issue:

```
import simtk.openmm as mm
from simtk import unit as u
import simtk.openmm.app as app

temperature = 300. * u.kelvin

pdb = app.PDBFile("/home/kyleb/src/openmm/openmm/wrappers/python/simtk/openmm/app/data/tip3p.pdb")
forcefield = app.ForceField('tip3p.xml')

system = forcefield.createSystem(pdb.topology, nonbondedMethod=app.PME, nonbondedCutoff=1.0*u.nanometers)
positions = pdb.positions


integrator = mm.CustomIntegrator(0.00001)
integrator.addGlobalVariable("U", 0.0)
integrator.addGlobalVariable("K", 0.0)
integrator.addUpdateContextState()
integrator.addComputePerDof("v", "v+0.0 * f")
integrator.addComputePerDof("x", "x+0.0 * f")
integrator.addComputeGlobal("U", "energy")
integrator.addComputeSum("K", "0.5*m*v*v")

context = mm.Context(system, integrator)
context.setPositions(positions)
context.setVelocitiesToTemperature(temperature)


integrator.step(1)

integrator.getGlobalVariableByName("U")
integrator.getGlobalVariableByName("K")

positions = context.getState(getPositions=True).getPositions()
state = context.getState(getEnergy=True)
energy = state.getPotentialEnergy() + state.getKineticEnergy()
energy, state.getPotentialEnergy(), state.getKineticEnergy()

integrator.getGlobalVariableByName("U")
integrator.getGlobalVariableByName("K")
```

**Comment by kyleabeauchamp** [comment]
Switching from PME to CutoffNonPeriodic eliminates the error, while going to CutoffPeriodic reduces the error to a few hundred kJ (below):

```  # CutoffPeriodic
In [26]: energy, state.getPotentialEnergy(), state.getKineticEnergy()
Out[26]: 
(Quantity(value=-29308.00949728462, unit=kilojoule/mole),
 Quantity(value=-35916.36008322212, unit=kilojoule/mole),
 Quantity(value=6608.3505859375, unit=kilojoule/mole))

In [27]: 

In [27]: integrator.getGlobalVariableByName("U")
Out[27]: -35761.65625
```

**Comment by peastman** [comment]
I see the problem.  The Ewald self energy and the long range dispersion correction aren't getting included.  I was trying to optimize this by having the integrator sum all the different energy components itself on the GPU.  But there are a few energy components that only ever get computed on the CPU.

Ok, I'll fix this.  There will be some performance cost, since it's going to require a full round trip to the CPU and back, but hopefully it won't be too bad.

**Comment by kyleabeauchamp** [comment]
Thanks Peter

**Comment by jchodera** [comment]
Good catch, @kyleabeauchamp and @peastman!

These terms are constant with box size, yes?  Would there be a way to compute these once on the CPU and cache them on the GPU when the `CustomIntegrator` is initialized?

Also, any idea how long this issue has been present?

**Comment by peastman** [comment]
Correct.  In fact, the Ewald self energy is simply a constant, independent even of the box size.

In principle these could be computed on the GPU as easily as the CPU.  But the problem is, the architecture leaves each Force object free to compute its energy wherever and however it wants.  In practice the CUDA platform does almost everything on the GPU, only leaving a couple of these trivial terms to the CPU.  But a plugin could do anything it wants.  The only really reliable way to do this is to go through the standard energy computation API.

I believe it's worked the current way since CustomIntegrator was first created.

**Comment by peastman** [comment]
Fixed by #883.
